### ISSUE TITLE
net.FQDN does not work as expected

### ISSUE THREAD

**Issue by sahroshan** [cue-lang/cue#2431]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version

cue version v0.5.0

go version go1.19.8
       -compiler gc
       -trimpath true
     CGO_ENABLED 0
          GOARCH amd64
            GOOS darwin
         GOAMD64 v1

</pre>

### Does this issue reproduce with the latest stable release?

yes

### What did you do?

1. create a cue file
```
import "net"

loki: {
	Endpoint: net.FQDN & string
}

```

2. create yaml file

```
loki:
  Endpoint: "https://abcd_ab.com"

```

3. export cue
```
$cue export abcd.cue abcd.yaml

{
    "loki": {
        "Endpoint": "https://abcd_ab.com"
    }
}
```
### What did you expect to see?

Cue should throw error as there is a _ in the domain name.

### What did you see instead?
No error was observed.

**Comment by slewiskelly** [comment]
I don't know if this is more of a bug in the [golang.org/x/net/idna](https://pkg.go.dev/golang.org/x/net/idna) package, but it seems like the `StrictDomainName` option can't be used without also using either of  the`MapForLookup` or `ValidateForRegistration` options.

I would propose replacing both `ValidateLabels` and `StrictDomainName` with `MapForLookup` as this sets both of these options also.

**Comment by slewiskelly** [comment]
https://review.gerrithub.io/c/cue-lang/cue/+/554858

**Comment by mvdan** [comment]
> I don't know if this is more of a bug in the [golang.org/x/net/idna](https://pkg.go.dev/golang.org/x/net/idna) package, but it seems like the `StrictDomainName` option can't be used without also using either of the`MapForLookup` or `ValidateForRegistration` options.

Given what the documentation says:

> StrictDomainName limits the set of permissible ASCII characters to those allowed in domain names as defined in [RFC 1034](https://rfc-editor.org/rfc/rfc1034.html) (A-Z, a-z, 0-9 and the hyphen). This is set by default for MapForLookup and ValidateForRegistration, but is only useful if ValidateLabels is set.

then I would expect our current code to work correctly - we do set `ValidateLabels`. I think this is a bug in the upstream package that we should raise there first. Either the upstream docs aren't very clear and we need to update our code, or upstream is buggy and we shouldn't change our code at all.

**Comment by mvdan** [comment]
Briefly spoke about this with Marcel - switching to `MapForLookup` per https://review.gerrithub.io/c/cue-lang/cue/+/554858 seems fine, it's a question of what the use cases for `FQDN` are. The current set of options we use are presumably less strict than "require the domain to be valid for lookups", even if they appear to be buggy.

I think we should aim to get `StrictDomainName` fixed upstream before we switch to `MapForLookup`. Upstream should get it fixed per the docs no matter what we end up doing here, and I also prefer to avoid breaking changes on our standard library unless we absolutely must do them.

**Comment by mvdan** [comment]
To reiterate - I'm happy with the change as proposed with two minor changes:

1) We should file a bug report upstream to flag that `StrictDomainName` does not behave as documented.
2) Our workaround to instead use `MapForLookup` should have a `// TODO` comment to be reverted once the upstream fix has been implemented and we can rely on its Go version.

@slewiskelly let me know if that sounds good and if you can file the upstream issue and update the CL. If you're not able to do that, I can always do it myself and pick up your change from its current state.

**Comment by vctrstrm** [comment]
Seeing same issue in 0.14 except it's more broken. - Any string, even string containing spaces passes through net.FQDN. Effectively, net.FQDN behaves exactly the same at `string & =~ ".+"`

**Comment by vctrstrm** [comment]
@slewiskelly i tested this patch on 0.14 and it seems to work correctly in all cases i could throw at it. https://github.com/cue-lang/cue/commit/de5274738efa209342f81d77ad8818ce07779523#diff-d731390ff0629a3dccc99b58773b2181472b4617ba74b15f9b473c86ffbab187

Did you want to resurrect that as a PR or should I?

**Comment by vctrstrm** [comment]
@mvdan I filed issue upstream https://github.com/golang/go/issues/74848

here's a go playground link replicating the behavior and workaround: https://go.dev/play/p/oCA7hxLYx4c

I'm new to the project so idk how you guys like to do PR's. Should i go ahead and make one?

**Comment by mvdan** [comment]
yes - as I mentioned above, I'm happy with a patch using that workaround as long as it includes a TODO linking to the upstream Go issue that you've created.
